cpp/src/arrow/scalar.cc

```diff
@@ -705,7 +713,9 @@ Result<std::shared_ptr<Scalar>> StructScalar::field(FieldRef ref) const {
 
 RunEndEncodedScalar::RunEndEncodedScalar(std::shared_ptr<Scalar> value,
                                          std::shared_ptr<DataType> type)
-    : Scalar{std::move(type), value->is_valid}, value{std::move(value)} {
+    : Scalar{std::move(type), value->is_valid},
+      ArraySpanFillFromScalarScratchSpace(run_end_type()->id()),
```

Comment:
Since this is implicitly `((RunEndEncodedScalar*)this)->run_end_type()` it will probably trigger the same issue eventually.
```suggestion
      ArraySpanFillFromScalarScratchSpace(*this->type),
```